File: proposed-projects/round-trip-experiment/reverse/20260909T140517Z_670261ce/20260910T081441Z_4899d5cf/ported_implementation/src/utils/errors/grammar-parse-error.ts
import { buildErrorPosition } from './build-error-position.js';

export const GRAMMAR_PARSER_ERROR_HEADER_MESSAGE = (reason: string): string =>
  `Failed to parse grammar: ${reason}`;

export class GrammarParseError extends Error {
  public readonly name = 'GrammarParseError';
  public readonly grammar: string;
  public readonly reason: string;
  public readonly pos: number;

  constructor(grammar: string, pos: number, reason: string) {
    super(
      [
        GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
        '',
        ...buildErrorPosition(grammar, pos),
      ].join('\n')
    );
    this.grammar = grammar;
    this.reason = reason;
    this.pos = pos;
  }
}

File: packages/gbnf-experiment/data/20260909T153339Z_9de9fd66/ported_implementation/gbnf/utils/errors/grammar_parse_error.py
from .build_error_position import build_error_position


def GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason: str) -> str:
    return f"Failed to parse grammar: {reason}"


class GrammarParseError(Exception):
    def __init__(self, grammar: str, pos: int, reason: str):
        message = "\n".join(
            [
                GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
                "",
                *build_error_position(grammar, pos),
            ]
        )
        super().__init__(message)
        self.name = "GrammarParseError"
        self.message = message
        self.grammar = grammar
        self.reason = reason
        self.pos = pos

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, GrammarParseError):
            return NotImplemented
        return (self.grammar, self.pos, self.reason) == (
            other.grammar,
            other.pos,
            other.reason,
        )

    def __hash__(self) -> int:
        return hash((self.grammar, self.pos, self.reason))
